### PR TITLE
Live debugger parsing, expression evaluation, sender, redaction and FFI

### PR DESCRIPTION
**Pull request by bwoebi**
Parsing goes through an intermediate step, which serde can simply deserialize to, handling which isn't possible by serde is then translated to working structures.

The Evaluator (feel free to clean up lifetime handling if possible. Spent far too much time fighting the borrow checker. Maybe my code is already perfect and I don't know it.) handles all evaluation of the expression language, iteration, error handling etc. and defers to some callbacks for actual evaluation of the values.

### CODE REVIEW

File: live-debugger-ffi/src/data.rs [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+    pub capture_snapshot: bool,
+    pub sampling_snapshots_per_second: u32,
+}
+
+impl<'a> From<&'a datadog_live_debugger::LogProbe> for LogProbe<'a> {
+    fn from(from: &'a datadog_live_debugger::LogProbe) -> Self {
+        LogProbe {
+            segments: &from.segments,
+            when: &from.when,
+            capture: &from.capture,
```
Comment (p-datadog) [inline live-debugger-ffi/src/data.rs:73]: What is the reason for these fields being pointers?

File: live-debugger-ffi/src/data.rs [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+    pub evaluate_at: EvaluateAt,
+    pub probe: ProbeType<'a>,
+}
+
+impl<'a> From<&'a datadog_live_debugger::Probe> for Probe<'a> {
+    fn from(from: &'a datadog_live_debugger::Probe) -> Self {
+        Probe {
+            id: from.id.as_str().into(),
+            version: from.version,
+            language: from.language.as_ref().map(|s| s.as_str().into()).into(),
```
Comment (p-datadog) [inline live-debugger-ffi/src/data.rs:210]: Is there a note/explanation somewhere for what this is for? Are there multiple string types being used?

File: live-debugger-ffi/src/sender.rs [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+    MaybeError::None
+}
+
+#[no_mangle]
+pub extern "C" fn ddog_live_debugger_send_raw_data(
+    handle: &mut SenderHandle,
+    data: OwnedCharSlice,
+) -> bool {
+    handle.channel.try_send(SendData::Wrapped(data)).is_err()
```
Comment (p-datadog) [inline live-debugger-ffi/src/sender.rs:126]: I tried using this function and it returns failure for me. The actual error is discarded on this line? How does one troubleshoot why this function is failing?

File: live-debugger/src/redacted_names.rs [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    }
+}
+
+pub fn is_redacted_name<'a, I: Into<&'a [u8]>>(name: I) -> bool {
+    fn invalid_char(c: u8) -> bool {
+        c == b'_' || c == b'-' || c == b'$' || c == b'@'
+    }
+    let name = name.into();
+    if name.len() > 20 {
+        return true; // short circuit for long names, assume them safe
```
Comment (p-datadog) [inline live-debugger/src/redacted_names.rs:122]: Are the names here potentially provided at runtime by the user/application? If so I don't see how it is valid to assume that names over 20 characters long are not subject to redaction.

File: live-debugger/src/sender.rs [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    match Client::builder()
+        .build(Connector::default())
+        .request(req)
+        .await
+    {
+        Ok(response) => {
+            if response.status().as_u16() >= 400 {
+                let body_bytes = hyper::body::to_bytes(response.into_body()).await?;
+                let response_body = String::from_utf8(body_bytes.to_vec()).unwrap_or_default();
+                anyhow::bail!("Server did not accept debugger payload: {response_body}");
```
Comment (p-datadog) [inline live-debugger/src/sender.rs:112]: Usually when HTTP errors are reported, they contain the status code along with the body.

File: live-debugger-ffi/src/data.rs (hunk 1 above)
Comment (bwoebi) [reply live-debugger-ffi/src/data.rs:73]: Avoiding unnecessary copies - this field is directly from an exportable rust representation and thus doesn't need a copy.

File: live-debugger-ffi/src/data.rs (hunk 2 above)
Comment (bwoebi) [reply live-debugger-ffi/src/data.rs:210]: Not sure what you are asking? Why I'm collecting the language? What this particular line of rust does? Or just asking what I'm doing to convert an Option<String> into an Option<CharSlice>?

File: live-debugger/src/sender.rs (hunk 5 above)
Comment (bwoebi) [reply live-debugger/src/sender.rs:112]: Agree, will enhance.

File: live-debugger/src/redacted_names.rs (hunk 4 above)
Comment (bwoebi) [reply live-debugger/src/redacted_names.rs:122]: They are, good point. I should keep track of the longest name.

File: live-debugger-ffi/src/sender.rs (hunk 3 above)
Comment (bwoebi) [reply live-debugger-ffi/src/sender.rs:126]: It returns true when there's an error. And false on success. I think that's the issue, i.e. missing an `!`.

But there's just one actual error behind this - when the channel is full, i.e. you're submitting log lines quicker than they can be consumed.

File: live-debugger-ffi/src/data.rs (hunk 2 above)
Comment (p-datadog) [reply live-debugger-ffi/src/data.rs:211]: Apologies for being unclear. I was asking why `language: from.language` would be insufficient, for my general education on how rust works.

Comment (bwoebi) [reply live-debugger-ffi/src/data.rs:211]: Ah, yeah, just the conversion of the value within the Option to something FFI understands.

**Review by ameske — APPROVED** [review]

File: live-debugger/src/redacted_names.rs [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+    let name = name.as_ref();
+    if name.len() > *ASSUMED_SAFE_NAME_LEN {
+        return true; // short circuit for long names, assume them safe
+    }
+    let mut copy = smallvec::SmallVec::<[u8; 21]>::with_capacity(name.len());
+    let mut i = 0;
+    while i < name.len() {
+        let mut c = name[i];
+        if !invalid_char(c) {
+            if c.is_ascii_uppercase() {
```
Comment (p-datadog) [inline live-debugger/src/redacted_names.rs:163]: Should this not handle utf-8 names?

Comment (bwoebi) [reply live-debugger/src/redacted_names.rs:163]: I did not see the need for utf-8 upper/lowercasing outside of the ascii alphabet. It comes at a certain processing cost... I.e. just lowercasing A-Z to a-z.

**Review by p-datadog — APPROVED** [review]

**Pull request merged**